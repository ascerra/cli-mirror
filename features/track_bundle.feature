Feature: track bundles
  The ec command line can track bundles and generate tracking data files

  Background:
    Given stub registry running

  Scenario:
    Given a tekton bundle image named "acceptance/bundle:tag" containing
      | Task     | task1     |
    When ec command is run with "track bundle --bundle ${REGISTRY}/acceptance/bundle:tag"
    Then the exit status should be 0
    Then the output should match the snapshot

  Scenario: Push data bundle to OCI registry
    Given a tekton bundle image named "acceptance/bundle:tag" containing
      | Task     | task1     |
    When ec command is run with "track bundle --bundle ${REGISTRY}/acceptance/bundle:tag --output oci:${REGISTRY}/tracked/bundle:tag"
    Then the exit status should be 0
    Then registry image "tracked/bundle:tag" should contain a layer with
    """
    ---
    task-bundles:
      ${REGISTRY}/acceptance/bundle:
        - digest: sha256:0af8c4f92f4b252b3ef0cbd712e7352196bc33a96c58b6e1d891b26e171deae8
          effective_on: "${TODAY_PLUS_30_DAYS}"
          tag: tag
    trusted_tasks:
      oci://${REGISTRY}/acceptance/bundle:tag:
        - effective_on: "${TODAY_PLUS_30_DAYS}"
          ref: sha256:0af8c4f92f4b252b3ef0cbd712e7352196bc33a96c58b6e1d891b26e171deae8

    """

  Scenario: Replace data bundle in OCI registry
    Given a tekton bundle image named "acceptance/bundle:1.0" containing
      | Task     | task1     |
    When ec command is run with "track bundle --bundle ${REGISTRY}/acceptance/bundle:1.0 --output oci:${REGISTRY}/tracked/bundle:tag"
    When a tekton bundle image named "acceptance/bundle:1.1" containing
      | Task     | task2     |
    When ec command is run with "track bundle --bundle ${REGISTRY}/acceptance/bundle:1.1 --input oci:${REGISTRY}/tracked/bundle:tag --replace"
    Then the exit status should be 0
    Then registry image "tracked/bundle:tag" should contain a layer with
    """
    ---
    task-bundles:
      ${REGISTRY}/acceptance/bundle:
        - digest: sha256:7af058b8a7adb24b74875411d625afbf90af6b4ed41b740606032edf1c4a0d1d
          effective_on: "${TODAY_PLUS_30_DAYS}"
          tag: "1.1"
        - digest: sha256:0af8c4f92f4b252b3ef0cbd712e7352196bc33a96c58b6e1d891b26e171deae8
          effective_on: "${TODAY_PLUS_30_DAYS}"
          tag: "1.0"
    trusted_tasks:
      oci://${REGISTRY}/acceptance/bundle:1.0:
        - effective_on: "${TODAY_PLUS_30_DAYS}"
          ref: sha256:0af8c4f92f4b252b3ef0cbd712e7352196bc33a96c58b6e1d891b26e171deae8
      oci://${REGISTRY}/acceptance/bundle:1.1:
        - effective_on: "${TODAY_PLUS_30_DAYS}"
          ref: sha256:7af058b8a7adb24b74875411d625afbf90af6b4ed41b740606032edf1c4a0d1d

    """

  Scenario: `ec track bundle` produced OPA bundle can be fetched via `conftest pull`
    Given a tekton bundle image named "acceptance/bundle:tag" containing
      | Task     | task1     |
    When ec command is run with "track bundle --bundle ${REGISTRY}/acceptance/bundle:tag --output oci:${REGISTRY}/tracked/bundle:tag"
    Then running conftest "pull oci://${REGISTRY}/tracked/bundle:tag" produces "policy/data/data/trusted_tekton_tasks.yml" containing:
    """
    ---
    task-bundles:
      ${REGISTRY}/acceptance/bundle:
        - digest: sha256:0af8c4f92f4b252b3ef0cbd712e7352196bc33a96c58b6e1d891b26e171deae8
          effective_on: "${TODAY_PLUS_30_DAYS}"
          tag: tag
    trusted_tasks:
      oci://${REGISTRY}/acceptance/bundle:tag:
        - effective_on: "${TODAY_PLUS_30_DAYS}"
          ref: sha256:0af8c4f92f4b252b3ef0cbd712e7352196bc33a96c58b6e1d891b26e171deae8

    """

  Scenario: Fresh tags
    Given a tekton bundle image named "acceptance/bundle:tag" containing
      | Task     | task1     |
      And a track bundle file named "${TMPDIR}/bundles.yaml" containing
    """
    ---
    task-bundles:
      ${REGISTRY}/acceptance/bundle:
        - digest: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}
          effective_on: 2006-01-02T15:04:05Z
          tag: tag
    """
      And a tekton bundle image named "acceptance/bundle:tag" containing
      | Task     | task1-updated     |
    When ec command is run with "track bundle --input ${TMPDIR}/bundles.yaml --bundle ${REGISTRY}/acceptance/bundle:tag"
    Then the exit status should be 0
    Then the output should match the snapshot

  Scenario: Pipeline definition is ignored from mixed bundle
    Given a tekton bundle image named "acceptance/bundle:tag" containing
      | Task     | task1     |
      | Pipeline | pipeline1 |
    When ec command is run with "track bundle --bundle ${REGISTRY}/acceptance/bundle:tag"
    Then the exit status should be 0
    Then the output should match the snapshot

  Scenario: Pipeline definition is ignored on its own
    Given a tekton bundle image named "acceptance/bundle:tag" containing
      | Pipeline | pipeline1 |
    When ec command is run with "track bundle --bundle ${REGISTRY}/acceptance/bundle:tag"
    Then the exit status should be 0
    Then the output should match the snapshot

  Scenario: Track git references
    When ec command is run with "track bundle --git git+https://github.com/redhat-appstudio/build-definitions.git//task/buildah/0.1/buildah.yaml@3672a457e3e89c0591369f609eba727b8e84108f"
    Then the exit status should be 0
    Then the output should match the snapshot

  Scenario: Track tekton-task alias
    Given a tekton bundle image named "acceptance/bundle:tag" containing
      | Task     | task1     |
    When ec command is run with "track tekton-task --bundle ${REGISTRY}/acceptance/bundle:tag --git git+https://github.com/redhat-appstudio/build-definitions.git//task/buildah/0.1/buildah.yaml@3672a457e3e89c0591369f609eba727b8e84108f"
    Then the exit status should be 0
    Then the output should match the snapshot

  Scenario: Track git references, append to existing
    Given a track bundle file named "${TMPDIR}/bundles.yaml" containing
    """
    ---
    trusted_tasks:
      git+https://forge.io/organization/repository.git//task/0.1/task.yaml:
        - effective_on: 2006-01-02T15:04:05Z
          ref: f0cacc1a
    """
    When ec command is run with "track tekton-task --input ${TMPDIR}/bundles.yaml --git git+https://forge.io/organization/repository.git//task/0.1/task.yaml@f0cacc1af00d"
    Then the exit status should be 0
    Then the output should match the snapshot

  Scenario: Track git references, with prune
    Given a track bundle file named "${TMPDIR}/bundles.yaml" containing
    """
    ---
    trusted_tasks:
      git+https://forge.io/organization/repository.git//task/0.1/task.yaml:
        - effective_on: 2006-01-02T15:04:05Z
          ref: f0cacc1a
    """
    When ec command is run with "track tekton-task --prune --input ${TMPDIR}/bundles.yaml --git git+https://forge.io/organization/repository.git//task/0.1/task.yaml@f0cacc1a"
    Then the exit status should be 0
    Then the output should match the snapshot
