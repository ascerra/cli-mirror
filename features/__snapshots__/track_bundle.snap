
[:stdout - 1]
/-/-/-/
task-bundles:
  ${REGISTRY}/acceptance/bundle:
    - digest: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}
      effective_on: "${TIMESTAMP}"
      tag: tag
trusted_tasks:
  oci://${REGISTRY}/acceptance/bundle:tag:
    - effective_on: "${TIMESTAMP}"
      ref: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}

---

[:stderr - 1]

---

[Fresh tags:stdout - 1]
/-/-/-/
task-bundles:
  ${REGISTRY}/acceptance/bundle:
    - digest: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}
      effective_on: "${TIMESTAMP}"
      tag: tag
    - digest: sha256:0af8c4f92f4b252b3ef0cbd712e7352196bc33a96c58b6e1d891b26e171deae8
      effective_on: "2006-01-02T15:04:05Z"
      expires_on: "${TIMESTAMP}"
      tag: tag
trusted_tasks:
  oci://${REGISTRY}/acceptance/bundle:tag:
    - effective_on: "${TIMESTAMP}"
      ref: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}
    - effective_on: "2006-01-02T15:04:05Z"
      expires_on: "${TIMESTAMP}"
      ref: sha256:0af8c4f92f4b252b3ef0cbd712e7352196bc33a96c58b6e1d891b26e171deae8

---

[Fresh tags:stderr - 1]

---

[Pipeline definition is ignored from mixed bundle:stdout - 1]
/-/-/-/
task-bundles:
  ${REGISTRY}/acceptance/bundle:
    - digest: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}
      effective_on: "${TIMESTAMP}"
      tag: tag
trusted_tasks:
  oci://${REGISTRY}/acceptance/bundle:tag:
    - effective_on: "${TIMESTAMP}"
      ref: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}

---

[Pipeline definition is ignored from mixed bundle:stderr - 1]

---

[Pipeline definition is ignored on its own:stdout - 1]
/-/-/-/
{}

---

[Pipeline definition is ignored on its own:stderr - 1]

---

[Track git references:stdout - 1]
/-/-/-/
trusted_tasks:
  git+https://github.com/redhat-appstudio/build-definitions.git//task/buildah/0.1/buildah.yaml:
    - effective_on: "${TIMESTAMP}"
      ref: 3672a457e3e89c0591369f609eba727b8e84108f

---

[Track git references:stderr - 1]

---

[track tekton-task alias:stdout - 1]
/-/-/-/
task-bundles:
  ${REGISTRY}/acceptance/bundle:
    - digest: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}
      effective_on: "${TIMESTAMP}"
      tag: tag
trusted_tasks:
  git+https://github.com/redhat-appstudio/build-definitions.git//task/buildah/0.1/buildah.yaml:
    - effective_on: "${TIMESTAMP}"
      ref: 3672a457e3e89c0591369f609eba727b8e84108f
  oci://${REGISTRY}/acceptance/bundle:tag:
    - effective_on: "${TIMESTAMP}"
      ref: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}

---

[track tekton-task alias:stderr - 1]

---

[Track git references:stdout - 2]
/-/-/-/
trusted_tasks:
  git+https://forge.io/organization/repository.git//task/0.1/task.yaml:
    - effective_on: "${TIMESTAMP}"
      ref: f0cacc1af00d

---

[Track git references:stderr - 2]

---

[Track tekton-task alias:stdout - 1]
/-/-/-/
task-bundles:
  ${REGISTRY}/acceptance/bundle:
    - digest: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}
      effective_on: "${TIMESTAMP}"
      tag: tag
trusted_tasks:
  git+https://github.com/redhat-appstudio/build-definitions.git//task/buildah/0.1/buildah.yaml:
    - effective_on: "${TIMESTAMP}"
      ref: 3672a457e3e89c0591369f609eba727b8e84108f
  oci://${REGISTRY}/acceptance/bundle:tag:
    - effective_on: "${TIMESTAMP}"
      ref: sha256:${REGISTRY_acceptance/bundle:tag_DIGEST}

---

[Track tekton-task alias:stderr - 1]

---

[Track git references, append to existing:stdout - 1]
/-/-/-/
trusted_tasks:
  git+https://forge.io/organization/repository.git//task/0.1/task.yaml:
    - effective_on: "${TIMESTAMP}"
      ref: f0cacc1af00d
    - effective_on: "2006-01-02T15:04:05Z"
      expires_on: "${TIMESTAMP}"
      ref: f0cacc1a

---

[Track git references, append to existing:stderr - 1]

---

[Track git references, with prune:stdout - 1]
/-/-/-/
trusted_tasks:
  git+https://forge.io/organization/repository.git//task/0.1/task.yaml:
    - effective_on: "2006-01-02T15:04:05Z"
      ref: f0cacc1a

---

[Track git references, with prune:stderr - 1]

---
